[Refactor] Some more C string removal

Geod24 opened the pull request:
A few more commits cleaning up strings while we wait on #8585 .

Geod24 committed:
Minor: Use Loc ctor instead of memberwise assignment

Geod24 committed:
Minor: Convert a `__gshared` to a `static immutable`

Geod24 committed:
imphint: Do not cast to string, use const(char)[]

Geod24 committed:
Privatize / unexport some manifest constants

Those constants are not exported to the C++ header,
so there is no point to have them extern(C++) __gshared
when the can just be immutable.

Geod24 committed:
doc: Modernize a manifest constant

Geod24 committed:
doc: Provide a slice overload of skipWhitespace

jacob-carlborg commented on src/dmd/doc.d:
```diff
@@ -1815,12 +1814,12 @@ extern (C++) int cmp(const(char)* stringz, const(void)* s, size_t slen)
 }
 
-extern (C++) int icmp(const(char)* stringz, const(void)* s, size_t slen)
+private bool icmp(const(char)* stringz, const(char)* s, size_t slen)
```
Isn’t the name a bit misleading now when it will only make a comparison.

Geod24 replied:
Definitely. Any suggestion ?

jacob-carlborg replied:
`equal`, `iequal`, `equali`, `equalCaseInsensitive`.

WalterBright commented on src/dmd/doc.d:
```diff
@@ -117,13 +117,12 @@ extern (C++) class Section
-                null
             ];
-            for (size_t i = 0; table[i]; i++)
+            foreach (const entry; table)
```
The `const` is redundant, as `table` is immutable.

WalterBright commented on src/dmd/doc.d:
```diff
@@ -1815,12 +1814,12 @@ extern (C++) int cmp(const(char)* stringz, const(void)* s, size_t slen)
-extern (C++) int icmp(const(char)* stringz, const(void)* s, size_t slen)
+private bool icmp(const(char)* stringz, const(char)* s, size_t slen)
 {
     size_t len1 = strlen(stringz);
```
Your fixes make all the first arguments to `icmp` a string, but then they get coerced to pointers by the parameter type, and then strlen is run on them here. Just commit to having string parameters to icmp.

WalterBright commented on src/dmd/doc.d:
```diff
@@ -1815,12 +1814,12 @@ extern (C++) int cmp(const(char)* stringz, const(void)* s, size_t slen)
-        return cast(int)(len1 - slen);
-    return Port.memicmp(stringz, cast(char*)s, slen);
+        return false;
+    return Port.memicmp(stringz, s, slen) == 0;
```
Give up on pointer-based memicmp and just write a loop.

WalterBright commented on src/dmd/errors.d:
```diff
@@ -73,10 +58,7 @@ extern (D) void error(Loc loc, const(char)* format, ...)
-    loc.filename = filename;
-    loc.linnum = linnum;
-    loc.charnum = charnum;
+    Loc loc = Loc(filename, linnum, charnum);
```
const

Geod24 committed:
utils: Introduce case insensitive slice comparison

This was previously used in doc, however as we move more towards
slice, the code could be simplified to the point it had to be removed.
Newly added `dmd.utils : iequals` will perform a case-insensitive string
comparison of two slices.

Geod24 commented:
Updated:
- Addressed Walter's comment
- Renamed/refactored `icmp` into `dmd.root.port.Port.iequals` and use slicing at call site

jacob-carlborg commented on src/dmd/root/port.d:
```diff
@@ -54,6 +54,34 @@ extern (C++) struct Port
+       Returns:
+         `true` if `s1 == s2` regardless of case
+     */
+    extern(D) static int iequals(const char[] s1, const char[] s2)
```
* Returns `int` instead of `bool`
* Is if worth mentioning in the documentation that this only works with ASCII (as far as I can tell)?

Geod24 commented:
Updated according to Jacob's comments:
```diff
diff --git a/src/dmd/root/port.d b/src/dmd/root/port.d
index ea52b617b..01ef74d53 100644
--- a/src/dmd/root/port.d
+++ b/src/dmd/root/port.d
@@ -57,6 +57,9 @@ extern (C++) struct Port
     /**
        Compare two slices for equality, in a case-insensitive way

+       Comparison is based on `char` and does not do decoding.
+       As a result, it's only really accurate for plain ASCII strings.
+
        Params:
          s1 = string to compare
          s2 = string to compare
@@ -64,7 +67,7 @@ extern (C++) struct Port
        Returns:
          `true` if `s1 == s2` regardless of case
      */
-    extern(D) static int iequals(const char[] s1, const char[] s2)
+    extern(D) static bool iequals(const(char)[] s1, const(char)[] s2)
     {
         if (s1.length != s2.length)
             return false;
```

WalterBright commented on src/dmd/doc.d:
```diff
@@ -117,13 +117,12 @@ extern (C++) class Section
+                if (Port.iequals(entry, name[0 .. namelen]))
                 {
-                    buf.printf("$(DDOC_%s ", table[i]);
+                    buf.printf("$(DDOC_%.*s ", entry.length, entry.ptr);
```
When using `.*s`, you must cast the `entry.length` parameter to `int`, because `entry.length` is a `size_t`.
Also, this change to `.*s` doesn't buy anything anyway, since the string is guaranteed to be 0 terminated.

Geod24 replied:
I reverted this bit to `.ptr`

Printing is the main issue I encountered when converting pointers to slices, and I think we need a more convenient way to print slices for it to move forward.

Geod24 commented on src/dmd/root/port.d:
```diff
@@ -54,6 +54,37 @@ extern (C++) struct Port
+       Returns:
+         `true` if `s1 == s2` regardless of case
+     */
+    extern(D) static bool iequals(const(char)[] s1, const(char)[] s2)
```
I was a bit surprised too, but decided to go with consistency.
Moved `iequals` to `utils`

Geod24 commented on src/dmd/dinifile.d:
```diff
@@ -351,7 +351,7 @@ void parseConfFile(StringTable* environment, const(char)* filename, const(char)*
                     if (!writeToEnv(environment, strdup(pn)))
                     {
-                        error(Loc(filename, lineNum, 0), "Use `NAME=value` syntax, not `%s`", pn);
+                        error(filename, lineNum, 0, "Use `NAME=value` syntax, not `%s`", pn);
```
You're right, I only considered the fact that this overload was used in only one place.
However changing this is a bit more involved and would trigger other changes, so I simply removed this commit from this PR and will do the changes in a separate PR.

Geod24 commented:
Updated:
```diff
diff --git a/src/dmd/imphint.d b/src/dmd/imphint.d
index 95f6319f0..57631750a 100644
--- a/src/dmd/imphint.d
+++ b/src/dmd/imphint.d
@@ -23,8 +23,8 @@ import dmd.utils;
  */
 extern (C++) const(char)* importHint(const(char)* s)
 {
-    if (auto ptr = s.toDString() in hints)
-        return (*ptr).ptr;
+    if (auto entry = s.toDString() in hints)
+        return entry.ptr;
     return null;
 }
diff --git a/src/dmd/doc.d b/src/dmd/doc.d
index 60b9b724d..44fc81782 100644
--- a/src/dmd/doc.d
+++ b/src/dmd/doc.d
@@ -1846,7 +1846,7 @@ extern (D) const(char)[] skipwhitespace(const(char)[] p)
         case ' ':
         case '\t':
         case '\n':
-            break;
+            continue;
         default:
             return p[idx .. $];
         }
diff --git a/src/dmd/link.d b/src/dmd/link.d
index 3334d2e0a..2d67e08e0 100644
--- a/src/dmd/link.d
+++ b/src/dmd/link.d
@@ -112,11 +112,11 @@ version (Posix)
     {
         version (OSX)
         {
-            static immutable nmeErrorMessage = "`__Dmain`, referenced from:";
+            static immutable(char*) nmeErrorMessage = "`__Dmain`, referenced from:";
         }
         else
         {
-            static immutable nmeErrorMessage = "undefined reference to `_Dmain`";
+            static immutable(char*) nmeErrorMessage = "undefined reference to `_Dmain`";
         }
         FILE* stream = fdopen(fd, "r");
         if (stream is null)
@@ -136,7 +136,7 @@ version (Posix)
             const(char)* lastSep = strrchr(buffer.ptr, '\n');
             if (lastSep)
                 buffer[(end = lastSep - &buffer[0])] = '\0';
-            if (strstr(&buffer[0], nmeErrorMessage.ptr))
+            if (strstr(&buffer[0], nmeErrorMessage))
                 nmeFound = true;
             if (lastSep)
                 buffer[end++] = '\n';
```

Geod24 commented:
Well this is green with auto-merge but dlang bot doesn't care, so merging manually
